main/streams/streams.c

```diff
@@ -1517,7 +1517,7 @@ PHPAPI zend_string *_php_stream_copy_to_mem(php_stream *src, size_t maxlen, int
 	ptr = ZSTR_VAL(result);
 
 	// TODO: Propagate error?
-	while ((ret = php_stream_read(src, ptr, max_len - len)) > 0){
+	while (!php_stream_eof(src) && (ret = php_stream_read(src, ptr, max_len - len)) > 0){
```

Comment:
If I am not mistaken, php_stream_eof() also generates some system calls (poll(), recv() with MSG_PEEK flag).

Reply:
Only if the `eof` flag has not yet been set:
https://github.com/php/php-src/blob/05023a281ddb62186fa47f51192ea51ba10f3a9b/main/streams/streams.c#L789-L791
But you are right, PHP's "xp_streams" implementation is rather clumsy. The way it is implemented, my change adds one system call **for streams** (replacing `recv()` with `poll()`+`recv()`); it still eliminates one extra `read()` system call for regular files!
The xp_streams code could be optimized easily. Anybody interested?

Reply:
Two system calls, when it's not at the EOF, we got `poll() + recv() + recv()`, and CHECK_LIVENESS call usually returns true.
php_stream is not only designed for files, it also works for sockets.

> The xp_streams code could be optimized easily. Anybody interested?

How? I can not understand.

Reply:
> Two system calls, when it's not at the EOF, we got `poll() + recv() + recv()`

Oh yes, you are right. I thought about the "eof" case, but forgot about the "not-eof".

> How? I can not understand.

By omitting the `poll()` on operating systems where `MSG_DONTWAIT` is available.

Reply:
> I always assume that doing so might cause a BC break.

How?

> In addition, this can also be applied when the stream is in non-blocking mode.

You mean on operating systems which don't support `MSG_DONTWAIT`? (i.e. Windows)